Add 051 release demo notebook

stuartarchibald opened the pull request:
As title.

stuartarchibald commented:
CC @gmarkall for CUDA link in.

CC @luk-f-a @StanczakDominik was wondering if either of you would like to do an initial review? No obligation/pressure :) Thanks!

StanczakDominik commented:
Sure, I'll take a look today :)

luk-f-a commented:
yes, no problem, I'll have a look today.

gmarkall commented:
The wording relating to the CUDA notebook looks good. When I click the link to the notebook on Binder, I get a "400: Bad Request" error though - is that to be expected at this stage prior to the merge of #21?

luk-f-a commented on notebooks/Numba_051_release_demo.ipynb:
```diff
@@ -0,0 +1,564 @@
+    "        # To access a field, we can define a function that simply\n",
+    "        # return the field in jit-code.\n",
+    "        # The definition is shown later.\n",
+    "        return Fruit_get_kind(self)\n",
```
writing this as the usual `return self.kind` crashes Python. Is this a bug? If not, a note highlighting that it must be done via an `njit` function would be helpful.

stuartarchibald replied:
The function is required to act as an accessor to the instance of the representation in the compiled code.

stuartarchibald commented on notebooks/Numba_051_release_demo.ipynb:
```diff
@@ -0,0 +1,564 @@
+    "Numba 0.51.0 Release Demo\n",
+    "=======================\n",
+    "\n",
+    "This notebook contains a demonstration of new features present in the 0.51.0 release of Numba. Whilst release notes are produced as part of the [`CHANGE_LOG`](https://github.com/numba/numba/blob/release0.51/CHANGE_LOG), there's nothing like seeing code in action!\n",
```
Thanks for highlighting this, I think `nbsphinx` could well be useful here, wonder what the complexity of using it cross repo is? The notebooks (and in fact this whole repo) are wired up to binder here https://mybinder.org/v2/gh/numba/numba-examples/master and in release announcements etc the live notebooks are referenced, so it's pre-action or action depending on how it was accessed! :)

For reference, the 0.50 release announcement is here https://numba.discourse.group/t/numba-0-50-0-and-llvmlite-0-33-0-released/19 with binder link https://mybinder.org/v2/gh/numba/numba-examples/master?filepath=notebooks%2FNumba_050_release_demo.ipynb

stuartarchibald commented:
> The wording relating to the CUDA notebook looks good. When I click the link to the notebook on Binder, I get a "400: Bad Request" error though - is that to be expected at this stage prior to the merge of #21?

@gmarkall  Thanks, and yes, needs to be on `master/` I think for binder to pick it up.

stuartarchibald commented on notebooks/Numba_051_release_demo.ipynb:
```diff
@@ -0,0 +1,564 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Heterogeneously typed lists also carry their type information, including literal values, such that it's possible to dispatch based on their value types"
```
Just from jitted functions, there's no unboxing for heterogeneous lists.

stuartarchibald replied:
Added a note in the notebook about that.

stuartarchibald commented on notebooks/Numba_051_release_demo.ipynb:
```diff
@@ -0,0 +1,564 @@
+    "\n",
+    "> What can I use as a mutable structure that's also pass-by-reference?\n",
+    "\n",
+    "the answer is the new ``StructRef`` type (warning: this is experimental!), [documentation is here](http://numba.pydata.org/numba-doc/dev/extending/high-level.html#implementing-mutable-structures)."
```
Undetermined at present! It's all experimental.

stuartarchibald commented:
Thanks for the review @StanczakDominik @luk-f-a, think I've addressed all your comments, please ping if not. I also had to update the notebook a bit with respect to last minute bug fixes too!

sklam commented on notebooks/Numba_051_release_demo.ipynb:
```diff
@@ -0,0 +1,564 @@
+    "    mixed_type_list_error()\n",
+    "except errors.TypingError as e:\n",
+    "    # CANNOT MUTATE A LITERAL LIST!\n",
+    "    assert \"Cannot mutate a literal list\" in str(e)"
```
a print was added

sklam commented on notebooks/Numba_051_release_demo.ipynb:
```diff
@@ -0,0 +1,564 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Initial value capturing\n",
```
A note was added below

sklam commented:
Looks like everything is addressed. Merging.
